pkgs/shells/fish/default.nix

```diff
@@ -35,28 +35,60 @@ let
     #     source both, but source the more global configuration files earlier
     #     than the more local ones, so that more local configurations inherit
     #     from but override the more global locations.
-    
+
     if test -f /etc/fish/config.fish
       source /etc/fish/config.fish
     end
-    
+
     #                                                                             #
     ############### ↑ Nix hook for sourcing /etc/fish/config.fish ↑ ###############
   '';
 
+
+  nixEnvironmentSetupScriptPrefix = "@NIX_STORE@/../var/nix/profiles/default/etc/profile.d";
+  selectPosixPreInitScript = if hostPlatform.isDarwin then ''
+    # anchor for indentation
+        # Use `set -U nix_on_darwin_no_daemon` to tell Fish to use the single-user mode script instead
```

Comment:
Is an environment variable the right place to configure this?

Reply:
How can we discern whether Nix should be using a daemon at this point in the configuration? Ordinarily, the Nix user chooses whether to use a daemon or not by modifying their shell config to source one script or the other. They can't modify the config that gets rolled into `fish` inside the Nix store.

This seems like a pretty natural solution to me. Do you think we should read from a file instead?

Reply:
If that's how it happens normally that's fine—I just wanted to make sure it was the normal place to do it. The only other option I was thinking was reading from /etc/nix/nix.conf but I don't think that actually has info about whether or not a daemon is used.

Reply:
@therealpxc Sorry, I'm not really qualified to review your changes: I updated the fish package a few time but I only have a vague idea of how the environment initializing works.

I'm a bit confused because I have never had the issue described in #18946.
For what is worth I'm ok for improving the situation for non-NixOs users.

Reply:
@rnhmjoj The issue is something which can occur if sourcing the Nix setup script happens redundantly or too late in shell initialization. It can be tricky to get right, but this PR uses the same logic as NixOS, which works by setting up the Nix environment before Fish uses any config.

@cbarrett there actually is some config that goes in /etc/nix/nix.conf that we might be able to use. I know `nix-build-users` is normally unset for single-user mode, for example.